specification/sections/03-evaluation-context.md

```diff
@@ -128,4 +130,52 @@ The SDK implementation must run the `on context changed` handler on all register
 
 > When the `evaluation context` for a specific provider is set, the `on context changed` handler **MUST** only run on the associated provider.
 
-The SDK implementation must run the `on context changed` handler only on the provider that is scoped to the mutated `evaluation context`.
\ No newline at end of file
+The SDK implementation must run the `on context changed` handler only on the provider that is scoped to the mutated `evaluation context`.
+
+### 3.3 Context Propagation
+
+[![experimental](https://img.shields.io/static/v1?label=Status&message=experimental&color=orange)](https://github.com/open-feature/spec/tree/main/specification#experimental)
+
+#### Condition 3.3.1
+
+> The implementation uses the dynamic-context paradigm.
+
+see: [dynamic-context paradigm](../glossary.md#dynamic-context-paradigm)
+
+##### Conditional Requirement 3.3.1.1
+
+> The API **MUST** have a method for setting a `transaction context propagator`.
+
+If there already is a `transaction context propagator`, it is replaced with the new one. 
+
+##### Conditional Requirement 3.3.1.2
+
+> The API **MUST** have a method for setting the `evaluation context` of the `transaction context propagator`.
+
+If a `transaction context propagator` is set, this `evaluation context` must be supplied to it and so will be available during the current transaction.
+This method can e.g. be used in a request handler to add request specific information to the `evaluation context`.
+
+##### Conditional Requirement 3.3.1.3
+
+> A `transaction context propagator` **MUST** have a method for setting the `evaluation context` of the current transaction.
+
+A `transaction context propagator` is responsible for persisting context for the duration of a single transaction.
+Typically, a transaction context propagator will propagate the context using a language-specific carrier such as [`ThreadLocal` (Java)](https://docs.oracle.com/javase/8/docs/api/java/lang/ThreadLocal.html), [`async hooks` (Node.js)](https://nodejs.org/api/async_hooks.html), [`Context` (Go)](https://pkg.go.dev/context) or another similar mechanism.
```

Comment:
Please add a code example.

Reply:
hi @lukas-reining,
if we look on real use-case consumer example in high level:

```
public isFeatureEnabled(HttpRequest request, String feature) {
        session = getSession(request);
        user = getUser(session);
        email = getEmail(request, session, user);
        evaluationContextMap.put(USER_EMAIL_PROPERTY, email);
        value = OpenFeatureAPI.getInstance().getClient().getBooleanValue(key, defaultValue, evaluationContext);
```

how will this spec change will handle this use case ?
Can the OpenFeature provider example treat this use case ?
Can you also provide high-level code example of consumer code that can match the consumer code example ?
Thanks

Reply:
What do you mean by this @liran2000? I think I do not really get the question.
In this case it seems that the feature flag is evaluated directly in the request handler is that right? The context propagation is meant for the case where you have information available at one place and want to have it available in all other places in this transaction. Like for example propagating a header value for use in repository (even if you may would not like to do this)

Reply:
@liran2000 here's an example in the OpenFeature playground app showing transaction context in action.

https://github.com/open-feature/playground/blob/main/packages/app/src/app/transaction-context.middleware.ts

The advantage is that you can set requested scoped context like a user email address in middleware and have it automatically be included during flag evaluation.

Reply:
@liran2000 this is how I see context propagation with regard to your example,

consider setting user email at the request handler implementation in Javaish  pseudocode,

```
    public void handleRequest(HttpRequest request){
        session = getSession(request);
        user = getUser(session);
        email = getEmail(request, session, user);

        evaluationContextMap.put(USER_EMAIL email);

        // set transaction context
        OpenFeatureAPI.setTransactionContext(new ImmutableContext(evaluationContextMap))
      
        .... rest of the logic
    }
 ```
 
 Once the context is set, any feature flag evaluation inside the current transaction will observe the `USER_EMAIL` property and merge it with other contexts (ex:- evaluation invocation context).
 
 For example, your example reduced to one line as email is already set in context,
 
 ```
  public isFeatureEnabled(HttpRequest request, String feature) { 
       ....
        OpenFeatureAPI.getInstance().getClient().getBooleanValue(key, defaultValue, invocationCtx);
    }
```

Reply:
Thank you all

Reply:
Please resolve this thread once you've had a chance to add a code snippet.

Reply:
added an example